net-analyzer/nuttcp/files/nuttcp.initd

```diff
@@ -0,0 +1,16 @@
+#!/sbin/openrc-run
+# Copyright 1999-2018 Gentoo Foundation
+# Distributed under the terms of the GNU General Public License v2
+
+command="/usr/bin/nuttcp"
+command_background="yes"
+command_args="${NUTTCP_OPTS}"
+pidfile="/run/${RC_SVCNAME}.pid"
+
+depend() {
+	need net
+}
+
+start_post() {
+	pgrep -n "${RC_SVCNAME%%.*}" > "${pidfile}"
```

Comment:
Ok, another question. Why do you use pidfile at all? If the program itself doesn't write it and you need to scan for the program name anyway, doesn't OpenRC just work™ without the pidfile? I think it will just scan for the program and kill it by name on `stop`, wouldn't it?

Reply:
> If the program itself doesn't write it and you need to scan for the program name anyway, doesn't OpenRC just work™ without the pidfile?

Yes, I need to scan.

See what happens without start_post section:

ebuild-x64-nuttcp /usr/local/portage/net-analyzer/nuttcp # /etc/init.d/nuttcp start
 * Caching service dependencies ...                                                                                                                                                                                        [ ok ]
 * No permission to apply cgroup settings
 * Starting nuttcp ...                                                                                                                                                                                                     [ ok ]
ebuild-x64-nuttcp /usr/local/portage/net-analyzer/nuttcp # pgrep nuttcp
7020
ebuild-x64-nuttcp /usr/local/portage/net-analyzer/nuttcp # cat /run/nuttcp.pid 
7017
ebuild-x64-nuttcp /usr/local/portage/net-analyzer/nuttcp # /etc/init.d/nuttcp stop
 * No permission to apply cgroup settings
 * Stopping nuttcp ...
 * start-stop-daemon: no matching processes found                                                                                                                                                                          [ ok ]
ebuild-x64-nuttcp /usr/local/portage/net-analyzer/nuttcp # pgrep nuttcp
7020
ebuild-x64-nuttcp /usr/local/portage/net-analyzer/nuttcp/files # cat /run/nuttcp.pid
cat: /run/nuttcp.pid: No such file or directory

Reply:
But what if you remove `pidfile` altogether?

Reply:
It works for one instance of nuttcp.
When I am trying to launch another one it doesn't work. Because it stops all instances of nuttcp regardless what instance I am trying to stop (no pidfiles - no way to distinguish one instance from another).

Or support for multiple instances for init is undesirable? And users who need them should setup it themselves?

Reply:
How useful are multiple instances? What I'm worried about is that things are going to be racey when multiple instances are enabled and openrc starts them in parallel. In that case it's possible that if both are started before `start_post` is executed, both instances will get the PID of one of them recorded.

Reply:
I use them widely.
I have production systems with more than a hundred of it.
```
backend spiderx # rc-update | grep ntripserver | wc -l
190
```
For me, "multiple instances" is a neat way to run various network related software on various ports, with various configuration options, and so on. But this is for me, and it is not a problem for me to create such inits. So if such support is undesirable I will stop to provide it in "my ebuilds" ( and will use it on my own).

> What I'm worried about is that things are going to be racey when multiple instances are enabled and openrc starts them in parallel. In that case it's possible that if both are started before start_post is executed, both instances will get the PID of one of them recorded.

I have never run into a problem with such approach (start_post) after more than a 6-7 years of openrc usage.

Reply:
Did you ever check that all those pidfiles have unique PID values? You won't notice the problem unless you try to stop one of them.

Reply:
Of course

ebuild-x64-nuttcp /usr/local # cat /run/nuttcp.pid 
1840
ebuild-x64-nuttcp /usr/local # cat /run/nuttcp.other.pid 
1863
ebuild-x64-nuttcp /usr/local # pgrep nuttcp
1840
1863

Reply:
I was talking about that system with 'hundreds of them', with the assumption that they're all started at boot, and parallel boot is enabled.

Reply:
backend spiderx # cat /run/ntripserver.*.pid > 1.txt
backend spiderx # pgrep ntripserver > 2.txt
backend spiderx # diff 1.txt 2.txt 
backend spiderx # 

They're all started at boot, in default runlevel.

And no, I don't use a parallel boot here.

backend spiderx # grep paralle /etc/rc.conf 
# in parallel for a slight speed improvement. When running in parallel we
# WARNING: whilst we have improved parallel, it can still potentially lock
#rc_parallel="NO"

I probably could clone this VM and try to enable parallel boot and see what will happen.
But even if I will run into a problem, a parallel boot is not enabled by default in gentoo. In other words, if this is a reason why we should not support multiple instances in openrc inits, I'd rather disable parallel boot than I refuse to use multiple instances.

Reply:
> I probably could clone this VM and try to enable parallel boot and see what will happen.

Looks like it works fine.

backend spiderx # cat /run/ntripserver.*.pid > 1.txt
backend spiderx # pgrep ntripserver > 2.txt
backend spiderx # cat 1.txt | sort > 1a.txt
backend spiderx # cat 2.txt | sort > 2a.txt
backend spiderx # diff 1a.txt 2a.txt 
backend spiderx #